Allow configuring Gradle, AGP, various SDK and Kotlin versions for Espresso server

tinder-ktarasov opened the pull request:
Currently, appium-espresso-driver uses same Gradle, AGP, SDK, and Kotlin versions when building espresso-server for all apps. Such versions might be different from the versions used to build AUT. Such setup is unusual as androidTest APKs are usually built with the same SDKs as the main app. It can lead to various issues as Android building infrastructure is quite complex.
This PR proposes changes that will allow configuring those versions to keep them in sync with the versions in the main app.

To build espresso-server with custom versions **forceEspressoRebuild** capability should be set to true.
I didn't change anything regarding caching, so I suppose rebuilt APK is not cached and it will be rebuilt even when versions are same.

~~Unfortunately, it doesn't fix that Resource issue, I'll make a separate PR for it.~~

**New desired capabilities**
- **showGradleLog** (boolean) - show Gradle build log in the server log
- **espressoBuildConfig** (path as a string) - path to build configuration JSON

**Build configuration JSON keys**
Configuration example:
```json
{
  "toolsVersions": {
    "gradle": "5.1.1",
    "androidGradlePlugin": "3.4.2",
    "compileSdk": 28,
    "buildTools": "28.0.3",
    "minSdk": 18,
    "targetSdk": 28,
    "kotlin": "1.3.31"
  },
  "additionalAppDependencies": [
    "com.google.android.material:material:1.0.0"
  ]
}
```

Minimum supported versions here and default versions are current versions in espresso-server project (not changed in this PR)
- **toolsVersion** - versions of various tools and SDKs used to build espresso-server
    - **gradle** - version of Gradle to use for building espresso-server
    - **androidGradlePlugin** - version of Android Gradle Plugin
    - **buildTools** - Android SDK build tools version
    - **compile_sdk_version** - compileSdk in build.gradle
    - **minSdk** - minSdk in build.gradle
    - **targetSdk** - targetSdk in build.gradle
    - **kotlin** - Kotlin version
- **additionalAppDependencies** - array containing additional dependencies of the tested app that build tools should know about. Items of the array maps to 'implementation' lines in build.gradle. I don't know exact causes but adding 'com.google.android.material:material:1.0.0' here fixes resource related startup issues for some apps.

**Possible improvements (not implemented)**
* Cache espresso-server APKs and reuse them when versions are the same
* Building espresso-server will spawn Gradle daemon process that is not stopped after the build (and will be reused for next builds). Probably some would want to avoid it.

mykola-mokhnach commented on espresso-server/build.gradle:
```diff
@@ -1,7 +1,25 @@
+        Object propertyValue = project.getProperties().get(propertyName)
+        if (propertyValue == null) {
+            return defaultValue
+        } else {
```
how about using the [elvis](https://mrhaki.blogspot.com/2009/08/groovy-goodness-elvis-operator.html) operator instead?

tinder-ktarasov replied:
I don't think using elvis operator is good idea here, so I replaced this block with ternary operator.

mykola-mokhnach commented on espresso-server/build.gradle:
```diff
@@ -1,7 +1,25 @@
+    }
+
+    ext.kotlin_version = getStringProperty('kotlin_version', '1.3.31')
+    ext.agp_version = getStringProperty('agp_version', '3.4.2')
```
what `agp` means?

tinder-ktarasov replied:
Android Gradle plugin

mykola-mokhnach replied:
I'd prefer to use full name instead of abbreviation then. It's not very common

mykola-mokhnach commented on lib/espresso-runner.js:
```diff
@@ -9,13 +10,38 @@ import B from 'bluebird';
+ * @param targetBaseDir {string} directory to copy files to
+ */
+async function copyRecursively (sourceBaseDir, targetBaseDir) {
+  logger.debug(`copyRecursively(${sourceBaseDir}, ${targetBaseDir})`);
```
I'd prefer human-readable log string ;)

tinder-ktarasov replied:
Actually I forgot to remove this log line. Do you think it makes sense to leave it and make more readable?

mykola-mokhnach replied:
yes, everything that helps to track the flow is helpful

mykola-mokhnach commented on lib/espresso-runner.js:
```diff
@@ -9,13 +10,38 @@ import B from 'bluebird';
+    if (!isDirectory) {
+      await fs.copyFile(itemPath, targetPath);
+    } else {
+      mkdirp(targetPath);
```
await

mykola-mokhnach commented on lib/driver.js:
```diff
@@ -345,6 +345,8 @@ class EspressoDriver extends BaseDriver {
       appPackage: this.opts.appPackage,
       appActivity: this.opts.appActivity,
       forceEspressoRebuild: !!this.opts.forceEspressoRebuild,
+      espressoBuildConfig: this.opts.espressoBuildConfig || '',
```
is it necessary to default this variable to an empty string?

mykola-mokhnach commented on lib/espresso-runner.js:
```diff
@@ -107,6 +135,20 @@ class EspressoRunner {
+    let versions = {};
+    if (this.espressoBuildConfig) {
+      logger.info(`Using build configuration JSON from: '${this.espressoBuildConfig}'`);
+      versions = JSON.parse(await fs.readFile(this.espressoBuildConfig));
```
there should be a proper human-readable message here if the file content is not a valid JSON

mykola-mokhnach replied:
readFile returns buffer by default. Provide the actual encoding as the second parameter to turn the result to a string

mykola-mokhnach commented on lib/espresso-runner.js:
```diff
@@ -9,13 +10,38 @@ import B from 'bluebird';
+ * @param sourceBaseDir {string} directory to copy files from
+ * @param targetBaseDir {string} directory to copy files to
+ */
+async function copyRecursively (sourceBaseDir, targetBaseDir) {
```
consider moving this function to utils.js

mykola-mokhnach replied:
it would be nice to have an integration test for it

tinder-ktarasov replied:
How would you implement such a test? By stubbing fs functions?

mykola-mokhnach commented on lib/server-builder.js:
```diff
@@ -0,0 +1,154 @@
+const gradleLog = logger.getLogger('Gradle');
+
+class ServerBuilder {
+  server_versions
```
what is the purpose of this variable?

tinder-ktarasov replied:
this is just explicit property declaration for syntax highlighting purposes. I suppose Appium style is to not declare them, so I'll remove it and review my code for similar small issues.

mykola-mokhnach replied:
we simply try to avoid static variables. Moving it to module level could be a good idea. We could also use LRU cache to make sure the value is never going to be too large

mykola-mokhnach commented on lib/server-builder.js:
```diff
@@ -0,0 +1,154 @@
+    this.serverPath = args.serverPath;
+    this.showGradleLog = args.showGradleLog;
+
+    const versions_arg = args.versions || {};
```
Please use javascript naming style (camelCase)

mykola-mokhnach commented on lib/server-builder.js:
```diff
@@ -0,0 +1,154 @@
+    let cmd = `.${path.sep}gradlew`;
+    let args = [];
+
+    for (const key of VERSION_KEYS) {
```
we could use map/filter here

mykola-mokhnach commented on lib/server-builder.js:
```diff
@@ -0,0 +1,154 @@
+      detached: true,
+      stdio: ['ignore', 'pipe', 'pipe'],
+    });
+    gradlebuild._last_lines = [];
```
`lastLines`

mykola-mokhnach replied:
why should this variable be attached to gradlebuild ?

tinder-ktarasov replied:
Here I've just copied approach from xcuitest driver. I don't think there are any issues with having two separate variables.

mykola-mokhnach commented on lib/server-builder.js:
```diff
@@ -0,0 +1,154 @@
+    });
+    gradlebuild._last_lines = [];
+
+    const logGradleOutput = this.showGradleLog;
```
why do we need this variable?

mykola-mokhnach replied:
bump

mykola-mokhnach commented on lib/server-builder.js:
```diff
@@ -0,0 +1,154 @@
+    gradlebuild._last_lines = [];
+
+    const logGradleOutput = this.showGradleLog;
+    const logMsg = _.isBoolean(this.showGradleLog)
```
will it work as expected? I saw that !!is applied above to this option

mykola-mokhnach commented on lib/server-builder.js:
```diff
@@ -0,0 +1,154 @@
+      gradlebuild.on('exit', (code, signal) => {
+        log.error(`Gradle exited with code '${code}' and signal '${signal}'`);
+
+        gradlebuild.processExited = true;
```
why this is needed?

mykola-mokhnach commented on lib/server-builder.js:
```diff
@@ -0,0 +1,154 @@
+      ? `Output from Gradle ${this.showGradleLog ? 'will' : 'will not'} be logged`
+      : 'Output from Gradle will only be logged if any errors are present there';
+    log.debug(`${logMsg}. To change this, use 'showGradleLog' desired capability`);
+    gradlebuild.on('output', (stdout, stderr) => {
```
consider listening to `lines-${streamName}` or `stream-line` messages. See https://github.com/appium/node-teen_process/blob/e06cf827cd428541ff972654406d788481909575/lib/subprocess.js#L45 for more details

mykola-mokhnach commented on lib/server-builder.js:
```diff
@@ -0,0 +1,154 @@
+    const gradlebuild = new SubProcess(cmd, args, {
+      cwd: this.serverPath,
+      env: process.env,
+      detached: true,
```
any reason to make the process detached?

mykola-mokhnach commented on lib/server-builder.js:
```diff
@@ -0,0 +1,154 @@
+      `in directory '${this.serverPath}'`);
+    const gradlebuild = new SubProcess(cmd, args, {
+      cwd: this.serverPath,
+      env: process.env,
```
Isn't this the default setting?

mykola-mokhnach commented on lib/server-builder.js:
```diff
@@ -0,0 +1,154 @@
+      .join('\n');
+  }
+
+  runBuildProcess () {
```
async

mykola-mokhnach replied:
it will work if you add await to `new B`
async modifier makes it easier to understand the function returns a promise

tinder-ktarasov replied:
I've reworked this function to avoid `detach` and it is much more readable now, also no `new B` now.

mykola-mokhnach commented:
We will need the feature and all the properties documented properly. https://github.com/appium/appium/tree/master/docs/en/writing-running-appium/android would be a good place to keep the document.

Altogether, I like the implementation, good job.

tinder-ktarasov commented on lib/server-builder.js:
```diff
@@ -0,0 +1,154 @@
+  }
+
+  getCommand () {
+    let cmd = `.${path.sep}gradlew`;
```
I didn't check the driver as whole on Windows, but running `.\gradlew` in command-line interpreter for Android project works on Windows.

mykola-mokhnach replied:
you should make sure it works from `cmd` command line. I suppose in windows the script has even different name (with `.bat` extension)

tinder-ktarasov replied:
Good point. Even though `.\gradlew` worked for me from cmd shell, it doesn't work from SubProcess.

KazuCocoa commented:
Awesome!

And it's good to add a description about the caps in https://github.com/appium/appium/blob/master/docs/en/writing-running-appium/caps.md after finalising this change.

tinder-ktarasov commented:
I've done some additional testing and found that our app require additional dependency on `com.google.android.material:material:1.0.0` in espresso-server. So I'd like to propose this configuration format:
```
{
  "toolsVersions": {
    "gradle": "5.1.1",
    "androidGradlePlugin": "3.4.2",
    "compileSdk": 28,
    "buildTools": "28.0.3",
    "minSdk": 18,
    "targetSdk": 28,
    "kotlin": "1.3.31"
  }
  "additionalAppDependencies": [// maps to 'implementation' lines in build.gradle
    "com.google.android.material:material:1.0.0"
  ]
}
```
What do you think?

KazuCocoa commented:
ah, nice. thanks :)
Will it allow to `exclude` or `include` as `additionalAppDependencies`? Or simply just add dependencies?

(I think only adding them is enough so far.)

tinder-ktarasov commented:
@mykola-mokhnach, I've addressed most of your comments. So there are 2 things left to do:
- implementing `additionalAppDependencies` suggestion above
- test for `copyRecursively`. I need some help with it - what is the best way to approach it and probably some example on stubbing fs functions (if such stubs are needed)

mykola-mokhnach commented:
> @mykola-mokhnach, I've addressed most of your comments. So there are 2 things left to do:

Thanks, nicely done

> test for copyRecursively. I need some help with it - what is the best way to approach it and probably some example on stubbing fs functions (if such stubs are needed)

I'm talking about a simple functional test. Create a temporary folder, put some subfolder with files/folders there, then copy the content to another temp folder and verify the copied content is there (files and (sub)folders).

mykola-mokhnach commented on lib/server-builder.js:
```diff
@@ -0,0 +1,124 @@
+        serverVersions[key] = versionsArg[key];
+      }
+    }
+    this.server_versions = serverVersions;
```
Please use camelCase naming

mykola-mokhnach commented on lib/server-builder.js:
```diff
@@ -0,0 +1,124 @@
+  }
+
+  getCommand () {
+    let cmd = system.isWindows() ? 'gradlew.bat' : './gradlew';
```
👍

mykola-mokhnach replied:
const

mykola-mokhnach replied:
Also I assume that `./` prefix is not needed for gradlew

tinder-ktarasov replied:
As I understand it, `./` prefix makes sure that gradlew will be started from the passed working directory, not from somewhere else in PATH. I think this would be more clean than going without `./`.

mykola-mokhnach replied:
not really. for a normal command prompt adding `./` prefix requests to execute the file
For subprocess having `cwd` option set is enough though

tinder-ktarasov replied:
Unfortunately just `gradlew` doesn't work on my mac, so I'll keep it `./gradlew`. For some reason it tries to run `gradlew` from PATH if there is no `./`.

mykola-mokhnach replied:
hm, ok. Thanks for checking it

mykola-mokhnach commented on lib/server-builder.js:
```diff
@@ -0,0 +1,124 @@
+        const serverVersion = this.server_versions[key];
+        return serverVersion ? `-P${key}=${serverVersion}` : null;
+      })
+      .filter(serverVersion => serverVersion);
```
filter(Boolean)

mykola-mokhnach commented on lib/server-builder.js:
```diff
@@ -0,0 +1,124 @@
+    return propertiesContent.split('\n')
+      .map(function (line) {
+        if (line.startsWith(GRADLE_URL_PREFIX)) {
+          return GRADLE_URL_PREFIX + GRADLE_URL_TEMPLATE.replace('VERSION', version);
```
use 

```
`${GRADLE_URL_PREFIX}{GRADLE_URL_TEMPLATE.replace('VERSION', version)}`
```

mykola-mokhnach commented on lib/server-builder.js:
```diff
@@ -0,0 +1,124 @@
+    } catch (err) {
+      let msg = `Unable to build Espresso server - ${err}\n` +
+        `Gradle error message:${EOL}${buildLastLines}`;
+      log.error(msg);
```
there is `log.errorAndThrow` method for this purpose

mykola-mokhnach commented on test/unit/server-builder-specs.js:
```diff
@@ -0,0 +1,62 @@
+import chai from 'chai';
```
👍

mykola-mokhnach commented on test/unit/server-builder-specs.js:
```diff
@@ -0,0 +1,62 @@
+    it('should pass only specified versions as properties', function () {
+      const expected = {cmd: expected_cmd, args: ['-Pandroid_gradle_plugin_version=1.2.3', 'assembleAndroidTest']};
+      let serverBuilder = new ServerBuilder({versions: {android_gradle_plugin_version: '1.2.3'}});
+      serverBuilder.getCommand().should.deep.eql(expected);
```
eql already does deep comparison

mykola-mokhnach commented on lib/espresso-runner.js:
```diff
@@ -1,15 +1,17 @@
 
 
-const TEST_APK_PATH = path.resolve(__dirname, '..', '..', 'espresso-server', 'app', 'build', 'outputs', 'apk', 'androidTest', 'debug', 'app-debug-androidTest.apk');
+const TEST_SERVER_PATH = path.resolve(__dirname, '..', '..', 'espresso-server');
```
how about TEST_SERVER_ROOT?

tinder-ktarasov commented on test/functional/utils-specs.js:
```diff
@@ -0,0 +1,87 @@
+  function expectNotExist (dir) {
+    expect(function () {
+      fs.accessSync(dir, fs.constants.F_OK);
+    }).to.throw();
```
I wasn't able to write this check using `async/await`+`chai-as-promised`, so I had to you sync version of the function. It is ok for me, if someone can rewrite it to be async.

mykola-mokhnach replied:
use `promiseFunc.should.eventually.be.rejected/rejectedWith(/pattern/)`

tinder-ktarasov commented on test/functional/utils-specs.js:
```diff
@@ -0,0 +1,87 @@
+
+  afterEach(async function () {
+    await fs.rimraf(baseSrcDir);
+    await fs.rimraf(baseDestDir);
```
My understanding of `rimraf` docs is that this method doesn't fail when directory doesn't exist. Could someone confirm this?

mykola-mokhnach replied:
yes it does not

mykola-mokhnach commented on test/functional/utils-specs.js:
```diff
@@ -0,0 +1,87 @@
+  }
+
+  beforeEach(function () {
+    baseSrcDir = nodefs.mkdtempSync(path.join(os.tmpdir(), 'appium-src-'));
```
use tempDir.openDir from appium-support module

mykola-mokhnach commented on lib/espresso-runner.js:
```diff
@@ -1,15 +1,17 @@
 
-const TEST_APK_PATH = path.resolve(__dirname, '..', '..', 'espresso-server', 'app', 'build', 'outputs', 'apk', 'androidTest', 'debug', 'app-debug-androidTest.apk');
+const TEST_SERVER_ROOT = path.resolve(__dirname, '..', '..', 'espresso-server');
 const TEST_MANIFEST_PATH = path.resolve(__dirname, '..', '..', 'espresso-server', 'AndroidManifest-test.xml');
```
this values could be replaces with path.resolve(TEST_SERVER_ROOT, 'AndroidManifest-test.xml')

mykola-mokhnach commented on lib/server-builder.js:
```diff
@@ -0,0 +1,161 @@
+        serverVersions[key] = versionConfiguration[key];
+      }
+    }
+    this.serverVersions = serverVersions;
```
this could nicely use VERSION_KEYS.reduce ...

mykola-mokhnach commented on lib/server-builder.js:
```diff
@@ -0,0 +1,161 @@
+    if (this.serverVersions[GRADLE_VERSION_KEY]) {
+      await this.setGradleWrapperVersion(this.serverVersions[GRADLE_VERSION_KEY]);
+    }
+    await this.insertAdditionalAppDependencies(this.additionalAppDependencies);
```
should we check if this.additionalAppDependencies is not empty first?

tinder-ktarasov replied:
It is not required as current implementation should not break anything. But yes, that will remove some IO calls when they are not required.

mykola-mokhnach commented on lib/server-builder.js:
```diff
@@ -0,0 +1,161 @@
+  }
+
+  updateGradleDistUrl (propertiesContent, version) {
+    return propertiesContent.split('\n')
```
I would rather replace it by regexp

tinder-ktarasov replied:
Ok. Probably `/[\r\n]/`?

mykola-mokhnach commented on lib/server-builder.js:
```diff
@@ -0,0 +1,161 @@
+    const propertiesPath = path.resolve(this.serverPath, 'gradle', 'wrapper', 'gradle-wrapper.properties');
+    const originalProperties = await fs.readFile(propertiesPath, 'utf8');
+    const newProperties = this.updateGradleDistUrl(originalProperties, version);
+    await fs.writeFile(propertiesPath, newProperties, 'utf8');
```
should we check if newProperties is different from originalProperties first?

tinder-ktarasov replied:
This line will only be executed if some Gradle version was specified in a build configuration file. So I'm not sure such check is needed.

mykola-mokhnach commented on lib/server-builder.js:
```diff
@@ -0,0 +1,161 @@
+    const dependencyLines = additionalAppDependencies
+      .map(function (dependency) {
+        if (/[\s']/.test(dependency)) {
+          throw new Error(`Disallowed characters in an additional dependency: ${dependency}`);
```
we probably would like to show which chars are not acceptable in the error message

mykola-mokhnach commented on lib/server-builder.js:
```diff
@@ -0,0 +1,161 @@
+      })
+      .join('\n');
+
+    return configurationContent.split('\n')
```
again, would prefer replacement by regexp

mykola-mokhnach commented on lib/server-builder.js:
```diff
@@ -0,0 +1,161 @@
+      }
+    });
+
+    buildLastLines = [];
```
why it should be assigned again?

mykola-mokhnach commented on lib/server-builder.js:
```diff
@@ -0,0 +1,161 @@
+      await gradlebuild.join();
+    } catch (err) {
+      let msg = `Unable to build Espresso server - ${err}\n` +
+        `Gradle error message:${EOL}${buildLastLines}`;
```
should we really collect these lines if we have `showGradleLog`? I would just suggest setting it to true if it is not set and checking the server log.

tinder-ktarasov replied:
That will complicate debugging flaky issues with building. So I would still like to have output for build failures, at least by default.
Also I think XCUITest driver prints error log when similar capability is `false`.

mykola-mokhnach commented on lib/utils.js:
```diff
@@ -22,4 +24,28 @@ function qualifyActivityName (activityName, packageName) {
+      return false;
+    }
+
+    if (!isDirectory) {
```
nit: positive conditions are easier to read

mykola-mokhnach commented on test/functional/utils-specs.js:
```diff
@@ -0,0 +1,82 @@
+  let baseDestDir;
+
+  function expectNotExist (file) {
+    fs.access(file, fs.constants.F_OK).should.eventually.be.rejectedWith(/no such file/);
```
await should still be there

mykola-mokhnach commented on test/functional/utils-specs.js:
```diff
@@ -0,0 +1,82 @@
+  }
+
+  function expectCorrectFileContentIn (filepath) {
+    fs.readFile(filepath, 'utf8').should.eventually.eql('foobar');
```
same here

tinder-ktarasov commented:
Signed the CLA

mykola-mokhnach commented on lib/server-builder.js:
```diff
@@ -32,13 +32,17 @@ class ServerBuilder {
-      }
-    }
-    this.serverVersions = serverVersions;
+    this.serverVersions = Object.entries(versionConfiguration)
```
the _reduce function for lodash lib could make it all much shorter. See https://github.com/appium/appium-idb/blob/2dce71d2b21b5d6bf95fcd5bf74e4c6bef0549ea/lib/helpers.js#L56

mykola-mokhnach commented on lib/server-builder.js:
```diff
@@ -47,7 +51,14 @@ class ServerBuilder {
+    if (!Array.isArray(this.additionalAppDependencies)) {
+      throw new Error('additionalAppDependencies must be an array');
+    }
+    if (this.additionalAppDependencies.length > 0) {
```
you could use _.isEmpty lodash function

mykola-mokhnach commented on lib/server-builder.js:
```diff
@@ -75,7 +86,7 @@ class ServerBuilder {
 
   updateGradleDistUrl (propertiesContent, version) {
-    return propertiesContent.split('\n')
+    return propertiesContent.split(/\n/)
```
I was talking about 

```
propertiesContent.replace(new RegExp(`^${_escapeRegex(GRADLE_URL_PREFIX)}`, 'gm'), `$1${GRADLE_URL_TEMPLATE.replace('VERSION', version)}`)
```

mykola-mokhnach commented on lib/server-builder.js:
```diff
@@ -94,19 +105,16 @@ class ServerBuilder {
       .map(function (dependency) {
         if (/[\s']/.test(dependency)) {
-          throw new Error(`Disallowed characters in an additional dependency: ${dependency}`);
+          throw new Error(`Single quotes and space characters are disallowed in additional dependencies: ${dependency}`);
```
I would say `space` -> `whitespace`, since `\s` also matches tabs, line breaks, etc

mykola-mokhnach commented on lib/server-builder.js:
```diff
@@ -94,19 +105,16 @@ class ServerBuilder {
       .join('\n');
 
-    return configurationContent.split('\n')
+    return configurationContent.split(/\n/)
```
same as above

hunzai commented:
@mykola-mokhnach @tinder-ktarasov let me know if I can help testing, at least :)

mykola-mokhnach commented on lib/server-builder.js:
```diff
@@ -32,17 +33,14 @@ class ServerBuilder {
-      .reduce(function (acc, [key, value]) {
-        return {...acc, [key]: value};
-      }, {});
+    this.serverVersions = _.reduce(versionConfiguration, (acc, value, key) => {
```
👍

mykola-mokhnach commented on lib/server-builder.js:
```diff
@@ -32,17 +33,14 @@ class ServerBuilder {
-        return {...acc, [key]: value};
-      }, {});
+    this.serverVersions = _.reduce(versionConfiguration, (acc, value, key) => {
+      if (!VERSION_KEYS.includes(key)) {
```
how about making it positive?

mykola-mokhnach commented on lib/server-builder.js:
```diff
@@ -107,22 +100,17 @@ class ServerBuilder {
-        if (/[\s']/.test(dependency)) {
-          throw new Error(`Single quotes and space characters are disallowed in additional dependencies: ${dependency}`);
+        if (/[\s'\\$]/.test(dependency)) {
+          throw new Error(`Single quotes, dollar characters and whitespace characters are disallowed in additional dependencies: ${dependency}`);
```
this line seems to be too long

mykola-mokhnach replied:
I mean this particular line.
You can simply split it using `+`

mykola-mokhnach commented on lib/server-builder.js:
```diff
@@ -107,22 +100,17 @@ class ServerBuilder {
-        }
-      })
-      .join('\n');
+      .join('$1');
```
will it work this way?

tinder-ktarasov replied:
Yes, this uses new line characters from the capture group in the pattern. I think I was wrong that I didn't add a comment here.

mykola-mokhnach commented on lib/server-builder.js:
```diff
@@ -107,22 +100,17 @@ class ServerBuilder {
+      .join('$1');
+
+    return configurationContent.replace(
+      new RegExp(`(\\s*^\\s*)${_.escapeRegExp(GRADLE_DEPENDENCIES_PLACEHOLDER)}\\s*$`, 'gm'),
```
Cannot `(\\s*^\\s*)` be just `(^\\s*)` ?

tinder-ktarasov replied:
`(\\s*^\\s*)` will capture new line characters currently used in the file to make sure implementation lines are correctly added. I'll add comment about it.

mykola-mokhnach commented:
It is necessary to make the tests green though.

tinder-ktarasov commented:
@mykola-mokhnach it doesn't look like the failing test case can be affected by these changes. Could you confirm that the test case about pointers is passing on `master` branch and it is not flaky?

tinder-ktarasov commented:
@hunzai you could check if `additionalAppDependencies` part of build configuration (see top comment in this PR) fixes issues with your app. You will need `npm link` the driver from this branch and restart Appium server to do it.

mykola-mokhnach commented:
@tinder-ktarasov Yep, the test is flaky. Restarting the job usually helps to resolve the problem

mykola-mokhnach commented on lib/server-builder.js:
```diff
@@ -0,0 +1,160 @@
+      // if we have an error we want to output the logs
+      // otherwise the failure is inscrutible
+      // but do not log expected errors
+      const ignoreError = IGNORED_ERRORS.some((x) => line.includes(x));
```
Why do we keep IGNORED_ERRORS array empty?

tinder-ktarasov replied:
For possible future filtering. Removed it now.

mykola-mokhnach replied:
YAGNI ;)

mykola-mokhnach commented on lib/server-builder.js:
```diff
@@ -0,0 +1,160 @@
+      const ignoreError = IGNORED_ERRORS.some((x) => line.includes(x));
+
+      if (this.showGradleLog && !ignoreError) {
+        gradleLog.error(line);
```
are all strings in the Gradle log errors? I would rather print them at debug or info level

tinder-ktarasov replied:
Probably keeping STDERR lines as `error` (that would be actual errors and task names) and changing other to `info` would be better?

mykola-mokhnach replied:
`error` is still too much. We only mark critical things as error. I would mark them as warnings and normal messages as info

mykola-mokhnach commented:
@tinder-ktarasov It would be nice to document the feature properly. You could put the resulting document to https://github.com/appium/appium/tree/master/docs/en/writing-running-appium/android

dpgraham commented:
@tinder-ktarasov do you want an invite to the Appium Slack channel?

tinder-ktarasov commented:
@mykola-mokhnach @KazuCocoa @dpgraham, sorry for being absent. I've added docs about this feature - https://github.com/appium/appium/pull/13766
